### PR TITLE
Add (partial) autodiff support for nll_loss

### PR DESCRIPTION
**Pull request by asuhan**
Not ready yet, need some comments / help with this. It's good enough for https://github.com/pytorch/xla immediate goals (forward + backward trace fusion), but there are at least two issues with it:

1. If we don't allow it, `test/test_jit.py` fails to cover the change.
2. If we allow the weight to be set, running `test/test_jit.py TestJitGenerated.test_nn_nll_loss` fails with:

```
======================================================================
ERROR: test_nn_nll_loss (__main__.TestJitGenerated)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 10001, in do_test
    fn, f_args_variable, kwargs_variable, no_grad=no_grad)
  File "test/test_jit.py", line 9360, in check_against_reference
    outputs_test = self.runAndSaveRNG(func, recording_inputs, kwargs)
  File "test/test_jit.py", line 425, in runAndSaveRNG
    results = func(*inputs, **kwargs)
  File "test/test_jit.py", line 9298, in script_fn
    self.assertExportImport(CU.the_method.graph, tensors)
  File "test/test_jit.py", line 415, in assertExportImport
    self.assertExportImportModule(m, inputs)
  File "test/test_jit.py", line 419, in assertExportImportModule
    self.assertEqual(self.runAndSaveRNG(m.forward, inputs),
  File "test/test_jit.py", line 425, in runAndSaveRNG
    results = func(*inputs, **kwargs)
RuntimeError: 
arguments for call are not valid:
  
  for operator aten::nll_loss_backward(Tensor grad_output, Tensor self, Tensor target, Tensor? weight, int reduction, int ignore_index, Tensor total_weight, *, Tensor out) -> Tensor:
  expected a value of type Tensor for argument 'total_weight' but found bool
  <internally-created-node>
  ~ <--- HERE
  
  for operator aten::nll_loss_backward(Tensor grad_output, Tensor self, Tensor target, Tensor? weight, int reduction, int ignore_index, Tensor total_weight) -> Tensor:
  expected a value of type Tensor for argument 'total_weight' but found bool
  <internally-created-node>
  ~ <--- HERE
for call at:
<internally-created-node>
~ <--- HERE
```

### CODE REVIEW

**Comment by asuhan** [comment]
@ailzhang Do you mind taking a look / taking over this?

**Comment by asuhan** [comment]
Having autodiff support for NLL is one of the last few things (the other being https://github.com/pytorch/pytorch/issues/14255) preventing the complete switch from our internal fork of PyTorch to the GitHub XLA extension. Synchronizing between the two is error prone and we need to wrap up the switch as soon as possible.

What's in this PR is sufficient for us and we have test coverage for our integration, but we should have coverage in the PyTorch core as well. If we can have nll loss with the new autodiff infrastructure within a few days, that works for us. However, if fixing this PR would be the shorter path, we'd prefer that.

**Comment by ailzhang** [comment]
@asuhan  I see, yea unblocking the integration using this PR might be the fastest solution. It looks good to me if all CI tests pass, do we want to add some manual tests as what @soumith did in #13888 ?

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -442,6 +447,21 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
       JIT_ASSERT(tuple_outputs.size() == size_t(3));
       return {tuple_outputs[0], tuple_outputs[1], tuple_outputs[2], nullptr, nullptr, nullptr, nullptr, nullptr};
 
+    } else if (node->matches("aten::nll_loss(Tensor self, Tensor target, Tensor? weight, int reduction, int ignore_index) -> Tensor")) {
+      auto graph = node->owningGraph();
```
Comment (ezyang) [inline torch/csrc/jit/autodiff.cpp]: I feel like it would be more defensive to ASSERT here the check that you did in the is differentiable check.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**